modules/wss4j/src/org/apache/ws/security/saml/SAMLIssuerImpl.java

```diff
@@ -148,110 +146,94 @@ public SAMLAssertion newAssertion() { // throws Exception {
                     }
-                    return null;
+                } else {
+                    X509Data certElem = new X509Data(instanceDoc);
+                    certElem.addCertificate(certs[0]);
+                    ki.add(certElem);
                 }
-                Element keyInfoElement = ki.getElement();
-                keyInfoElement.setAttributeNS(WSConstants.XMLNS_NS, "xmlns:"
-                        + WSConstants.SIG_PREFIX, WSConstants.SIG_NS);
-
-                subject.setKeyInfo(ki);
-                // prepare to sign the SAML token
-                try {
-                    X509Certificate[] issuerCerts =
-                            issuerCrypto.getCertificates(issuerKeyName);
-
-                    String sigAlgo = XMLSignature.ALGO_ID_SIGNATURE_RSA;
-                    String pubKeyAlgo =
-                            issuerCerts[0].getPublicKey().getAlgorithm();
-                    log.debug("automatic sig algo detection: " + pubKeyAlgo);
-                    if (pubKeyAlgo.equalsIgnoreCase("DSA")) {
-                        sigAlgo = XMLSignature.ALGO_ID_SIGNATURE_DSA;
-                    }
-                    java.security.Key issuerPK =
-                            issuerCrypto.getPrivateKey(issuerKeyName,
-                                    issuerKeyPassword);
-                    sa.sign(sigAlgo, issuerPK, Arrays.asList(issuerCerts));
-                } catch (WSSecurityException ex) {
-                    if (log.isDebugEnabled()) {
-                        log.debug(ex.getMessage(), ex);
-                    }
-                    return null;
-                } catch (Exception ex) {
-                    if (log.isDebugEnabled()) {
-                        log.debug(ex.getMessage(), ex);
-                    }
-                    return null;
+            } catch (WSSecurityException ex) {
+                if (log.isDebugEnabled()) {
+                    log.debug(ex.getMessage(), ex);
+                }
+                return null;
+            } catch (XMLSecurityException ex) {
+                if (log.isDebugEnabled()) {
+                    log.debug(ex.getMessage(), ex);
                 }
+                return null;
             }
-        } catch (SAMLException ex) {
-            if (log.isDebugEnabled()) {
-                log.debug(ex.getMessage(), ex);
+            Element keyInfoElement = ki.getElement();
+            keyInfoElement.setAttributeNS(WSConstants.XMLNS_NS, "xmlns:"
+                    + WSConstants.SIG_PREFIX, WSConstants.SIG_NS);
+
+//                subject.setKeyInfo(ki);
+            // prepare to sign the SAML token
+            try {
+                X509Certificate[] issuerCerts =
+                        issuerCrypto.getCertificates(issuerKeyName);
+
+                String sigAlgo = XMLSignature.ALGO_ID_SIGNATURE_RSA;
+                String pubKeyAlgo =
+                        issuerCerts[0].getPublicKey().getAlgorithm();
```

Comment:
better check for size of array (and for null) to avoid index OOB

Reply:
fixed in 2f696f6746f53ec59ed149ea7753dbcb09f7b044